File: bindings/js/test/golden.mjs
import assert from "node:assert/strict";
import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { dirname, resolve } from "node:path";
import {
  EventLog,
  Normalization,
  PromotionCounter,
  absoluteDistance,
  adaptiveLadderDistance,
  carryRules,
  consensusResultFromSchema,
  customLadder,
  euclideanDistance,
  geometricLadder,
  ladderDistance,
  ladderFromSchema,
  ladderPair,
  ladderToSchema,
  normalizeTicks,
  smoothLadderDistance,
  smoothTickDistance,
  tickDistance,
  tickPair,
  tickVectorFromSchema,
  tickVectorToSchema,
  tier,
  tierFromSchema,
  tierToSchema,
  validateLadder,
  weightedConsensus,
} from "../src/index.js";

const root = resolve(dirname(fileURLToPath(import.meta.url)), "../../../crates/metricchrono-core");
const repoRoot = resolve(root, "../..");
const eps = 1e-12;
const fixture = readJson(resolve(root, "fixtures/binding_conformance.v1.json"));

assert.equal(fixture.metricchrono_schema, "binding_conformance.v1");

for (const row of readCsv("fixtures/golden_ticks.csv").slice(1)) {
  const t = tier(number(row[2]), number(row[3]), number(row[4]), number(row[5]));
  assertClose(row[0], tickDistance(number(row[1]), t), number(row[6]));
}

for (const row of readCsv("fixtures/golden_ladders.csv").slice(1)) {
  const ladder = geometricLadder(
    number(row[2]),
    number(row[3]),
    number(row[4]),
    Number.parseInt(row[5], 10),
    number(row[6]),
    number(row[7]),
  );
  const actual = ladderDistance(number(row[1]), ladder);
  const expected = row[8].split(";").map(number);
  assertVectorClose(row[0], actual, expected);
}

const ladders = buildLadders(fixture);

for (const testCase of fixture.tick_distance_cases) {
  const t = tierFromDoc(testCase.tier);
  assertClose(testCase.name, tickDistance(testCase.distance, t), testCase.expected);
}

for (const testCase of fixture.smooth_tick_distance_cases) {
  assertClose(
    testCase.name,
    smoothTickDistance(testCase.distance, tierFromDoc(testCase.tier), testCase.sharpness),
    testCase.expected,
  );
}

for (const testCase of fixture.smooth_ladder_distance_cases) {
  assertVectorClose(
    testCase.name,
    smoothLadderDistance(testCase.distance, ladders.get(testCase.ladder), testCase.sharpness),
    testCase.expected,
  );
}

for (const testCase of fixture.adaptive_ladder_distance_cases) {
  const actual = adaptiveLadderDistance(testCase.distance, ladders.get(testCase.ladder));
  assertVectorClose(testCase.name, actual.ticks, testCase.expected.ticks);
  assert.equal(actual.decision.evaluatedTiers, testCase.expected.decision.evaluated_tiers);
  assert.equal(actual.decision.firstInactiveTier, testCase.expected.decision.first_inactive_tier);
  assert.equal(actual.decision.stoppedEarly, testCase.expected.decision.stopped_early);
}

for (const testCase of fixture.weighted_consensus_cases) {
  assertVectorClose(
    testCase.name,
    weightedConsensus(testCase.vectors, testCase.weights),
    testCase.expected,
  );
}

for (const testCase of fixture.event_log_cases) {
  assertEventLogCase(testCase);
}

for (const testCase of fixture.metric_cases.euclidean_distance) {
  assertClose(testCase.name, euclideanDistance(testCase.a, testCase.b), testCase.expected);
}

for (const testCase of fixture.metric_cases.absolute_distance) {
  assertClose(testCase.name, absoluteDistance(testCase.a, testCase.b), testCase.expected);
}

for (const testCase of fixture.pair_cases.tick_pair) {
  assertClose(
    testCase.name,
    tickPair(testCase.a, testCase.b, metricFromName(testCase.metric), tierFromDoc(testCase.tier)),
    testCase.expected,
  );
}

for (const testCase of fixture.pair_cases.ladder_pair) {
  assertVectorClose(
    testCase.name,
    ladderPair(testCase.a, testCase.b, metricFromName(testCase.metric), ladders.get(testCase.ladder)),
    testCase.expected,
  );
}

for (const testCase of fixture.normalize_ticks_cases) {
  assertVectorClose(
    testCase.name,
    normalizeTicks(testCase.input, normalizationMode(testCase.mode)),
    testCase.expected,
  );
}

for (const testCase of fixture.carry_rules_cases) {
  assert.deepEqual(carryRules(testCase.epsilons), testCase.expected, testCase.name);
}

for (const testCase of fixture.promotion_counter_cases) {
  const counter = new PromotionCounter(testCase.quotas);
  assert.deepEqual(counter.quotas, testCase.quotas, `${testCase.name}/quotas`);
  for (const step of testCase.steps) {
    const actual = step.event_flags === null ? counter.step() : counter.step(step.event_flags);
    assert.deepEqual(actual, step.promoted, `${testCase.name}/promoted`);
    assert.deepEqual(counter.counters, step.counters, `${testCase.name}/counters`);
  }
  counter.reset();
  assert.deepEqual(counter.counters, testCase.after_reset_counters, `${testCase.name}/reset`);
}

assertRejections(fixture.rejections, fixture.event_log_cases[0]);
assertSchemaRoundTrip();

function buildLadders(source) {
  const out = new Map();
  for (const testCase of source.ladders) {
    const expected = testCase.tiers.map(tierFromDoc);
    let actual;
    if (testCase.kind === "geometric") {
      const params = testCase.params;
      actual = geometricLadder(
        params.epsilon0,
        params.delta0,
        params.ratio,
        params.tiers,
        params.p,
        params.epsilon_ref,
      );
    } else if (testCase.kind === "custom") {
      actual = customLadder(expected);
    } else {
      throw new Error(`unknown ladder kind ${testCase.kind}`);
    }

    validateLadder(actual);
    assert.equal(actual.length, expected.length, `${testCase.name}/tiers`);
    for (let index = 0; index < actual.length; index += 1) {
      assertClose(`${testCase.name}[${index}].epsilon`, actual[index].epsilon, expected[index].epsilon);
      assertClose(`${testCase.name}[${index}].delta`, actual[index].delta, expected[index].delta);
      assertClose(`${testCase.name}[${index}].p`, actual[index].p, expected[index].p);
      assertClose(
        `${testCase.name}[${index}].epsilonRef`,
        actual[index].epsilonRef,
        expected[index].epsilonRef,
      );
    }
    for (const distanceCase of testCase.distances) {
      assertVectorClose(
        `${testCase.name}/ladderDistance`,
        ladderDistance(distanceCase.distance, actual),
        distanceCase.expected,
      );
    }
    out.set(testCase.name, actual);
  }
  return out;
}

function assertEventLogCase(testCase) {
  const log = new EventLog(testCase.tier_count);
  assert.equal(log.length === 0, testCase.is_empty_before_append, `${testCase.name}/isEmptyBefore`);
  for (const record of testCase.append_records) {
    assert.equal(log.append(record.state_id, record.ticks), record.expected_index);
  }
  assert.equal(log.length, testCase.expected_len, `${testCase.name}/length`);
  assert.equal(log.length === 0, testCase.is_empty_after_append, `${testCase.name}/isEmptyAfter`);
  assert.equal(log.tierCount, testCase.tier_count, `${testCase.name}/tierCount`);

  for (const expected of testCase.records) {
    const actual = log.records[expected.index];
    assert.equal(actual.stateId, expected.state_id, `${testCase.name}/recordState`);
    assertVectorClose(`${testCase.name}/record[${expected.index}]`, actual.ticks, expected.ticks);
  }

  for (const expected of testCase.first_events) {
    assert.equal(log.firstEvent(expected.tier), expected.expected, `${testCase.name}/firstEvent`);
  }

  for (const expected of testCase.next_events) {
    assert.equal(
      log.nextEvent(expected.index, expected.tier),
      expected.expected,
      `${testCase.name}/nextEvent`,
    );
  }

  for (const summary of testCase.compact_summaries) {
    const actual = log.compactSummary(summary.tier);
    assert.equal(actual.length, summary.expected.length, `${testCase.name}/summaryLength`);
    for (let index = 0; index < actual.length; index += 1) {
      assert.equal(actual[index].index, summary.expected[index].index);
      assert.equal(actual[index].stateId, summary.expected[index].state_id);
      assertClose("compact summary tick", actual[index].tick, summary.expected[index].tick);
    }
  }
}

function assertRejections(rejections, eventCase) {
  for (const testCase of rejections.invalid_tiers) {
    assert.throws(
      () => tier(testCase.epsilon, testCase.delta, testCase.p, testCase.epsilon_ref),
      undefined,
      testCase.name,
    );
  }

  for (const testCase of rejections.unknown_schema_documents) {
    if (testCase.kind === "tier") {
      assert.throws(() => tierFromSchema(testCase.document), undefined, testCase.name);
    } else {
      throw new Error(`unknown schema rejection kind ${testCase.kind}`);
    }
  }

  const log = new EventLog(eventCase.tier_count);
  for (const record of eventCase.append_records) {
    log.append(record.state_id, record.ticks);
  }
  for (const testCase of rejections.event_log_out_of_range) {
    if (testCase.operation === "record") {
      assert.throws(() => {
        const record = log.records[testCase.index];
        if (record === undefined) {
          throw new Error("event log index is out of bounds");
        }
      }, undefined, testCase.name);
    } else if (testCase.operation === "next_event") {
      assert.throws(() => log.nextEvent(testCase.index, testCase.tier), undefined, testCase.name);
    } else if (testCase.operation === "first_event") {
      assert.throws(() => log.firstEvent(testCase.tier), undefined, testCase.name);
    } else if (testCase.operation === "compact_summary") {
      assert.throws(() => log.compactSummary(testCase.tier), undefined, testCase.name);
    } else {
      throw new Error(`unknown event log rejection operation ${testCase.operation}`);
    }
  }
}

function assertSchemaRoundTrip() {
  const tierDoc = {
    metricchrono_schema: "tier.v1",
    epsilon: 0.03,
    delta: 0.1,
    p: 0.0,
    epsilon_ref: 1.0,
  };
  const schemaTier = tierFromSchema(tierDoc);
  assert.deepEqual(tierToSchema(schemaTier), tierDoc);

  const ladderDoc = readJson(resolve(repoRoot, "tests/golden/ladder.v1.json"));
  const schemaLadder = ladderFromSchema(ladderDoc);
  assert.deepEqual(ladderDistance(1.0, schemaLadder), [10, 4, 2]);
  assert.deepEqual(ladderToSchema(schemaLadder), ladderDoc);

  const tickDoc = readJson(resolve(repoRoot, "tests/golden/tick_vector.v1.json"));
  const schemaTicks = tickVectorFromSchema(tickDoc);
  assert.deepEqual(schemaTicks, [10, 4, 2]);
  assert.deepEqual(tickVectorToSchema(schemaTicks), tickDoc);

  const consensusDoc = readJson(resolve(repoRoot, "tests/golden/consensus_result.v1.json"));
  assert.deepEqual(consensusResultFromSchema(consensusDoc), consensusDoc);

  assert.throws(() => ladderFromSchema({ ...ladderDoc, metricchrono_schema: "ladder.v2" }));
}

function metricFromName(name) {
  if (name === "euclidean") {
    return euclideanDistance;
  }
  if (name === "absolute") {
    return absoluteDistance;
  }
  if (name === "max_abs") {
    return maxAbsDistance;
  }
  throw new Error(`unknown metric ${name}`);
}

function maxAbsDistance(a, b) {
  if (a.length !== b.length) {
    return Number.NaN;
  }
  return a.map((value, index) => Math.abs(value - b[index])).reduce((left, right) => Math.max(left, right), 0);
}

function normalizationMode(mode) {
  if (mode === "none") {
    return Normalization.None;
  }
  if (mode === "unit_max") {
    return Normalization.UnitMax;
  }
  if (mode === "tanh") {
    return Normalization.Tanh;
  }
  throw new Error(`unknown normalization ${mode}`);
}

function tierFromDoc(document) {
  return tier(document.epsilon, document.delta, document.p, document.epsilon_ref);
}

function readCsv(path) {
  return readFileSync(resolve(root, path), "utf8")
    .trim()
    .split("\n")
    .map((line) => line.split(","));
}

function readJson(path) {
  return JSON.parse(readFileSync(path, "utf8"));
}

function number(value) {
  return Number.parseFloat(value);
}

function assertVectorClose(name, actual, expected) {
  assert.equal(actual.length, expected.length, `${name}: length mismatch`);
  for (let index = 0; index < actual.length; index += 1) {
    assertClose(`${name}[${index}]`, actual[index], expected[index]);
  }
}

function assertClose(name, actual, expected) {
  assert.ok(Math.abs(actual - expected) <= eps, `${name}: expected ${expected}, got ${actual}`);
}

File: bindings/python/tests/golden.py
from __future__ import annotations

import csv
import json
import math
from pathlib import Path
from typing import Any

import metricchrono as mc

ROOT = Path(__file__).resolve().parents[3] / "crates/metricchrono-core"
REPO_ROOT = ROOT.parents[1]
CONFORMANCE = ROOT / "fixtures/binding_conformance.v1.json"
EPS = 1e-12


def assert_close(name: str, actual: float, expected: float, eps: float = EPS) -> None:
    if abs(actual - expected) > eps:
        raise AssertionError(f"{name}: expected {expected}, got {actual}")


def assert_vector_close(
    name: str,
    actual: list[float] | tuple[float, ...],
    expected: list[float],
    eps: float = EPS,
) -> None:
    if len(actual) != len(expected):
        raise AssertionError(f"{name}: length mismatch, expected {len(expected)}, got {len(actual)}")
    for index, (left, right) in enumerate(zip(actual, expected)):
        assert_close(f"{name}[{index}]", float(left), float(right), eps)


def load_conformance() -> dict[str, Any]:
    with CONFORMANCE.open() as handle:
        return json.load(handle)


def tier_from_doc(document: dict[str, Any]) -> mc.Tier:
    return mc.Tier(
        float(document["epsilon"]),
        float(document["delta"]),
        float(document["p"]),
        float(document["epsilon_ref"]),
    )


def ladder_map(fixture: dict[str, Any]) -> dict[str, list[mc.Tier]]:
    ladders: dict[str, list[mc.Tier]] = {}
    for case in fixture["ladders"]:
        expected = [tier_from_doc(tier) for tier in case["tiers"]]
        if case["kind"] == "geometric":
            params = case["params"]
            actual = mc.geometric_ladder(
                params["epsilon0"],
                params["delta0"],
                params["ratio"],
                params["tiers"],
                params["p"],
                params["epsilon_ref"],
            )
            object_ladder = mc.Ladder.geometric(
                params["epsilon0"],
                params["delta0"],
                params["ratio"],
                params["tiers"],
                params["p"],
                params["epsilon_ref"],
            )
        elif case["kind"] == "custom":
            actual = mc.custom_ladder(expected)
            object_ladder = mc.Ladder(expected)
        else:
            raise AssertionError(f"unknown ladder kind {case['kind']}")

        mc.validate_ladder(actual)
        if len(actual) != len(expected):
            raise AssertionError(f"{case['name']}: tier length mismatch")
        for index, (left, right) in enumerate(zip(actual, expected)):
            assert_close(f"{case['name']}[{index}].epsilon", left.epsilon, right.epsilon)
            assert_close(f"{case['name']}[{index}].delta", left.delta, right.delta)
            assert_close(f"{case['name']}[{index}].p", left.p, right.p)
            assert_close(f"{case['name']}[{index}].epsilon_ref", left.epsilon_ref, right.epsilon_ref)

        for distance_case in case["distances"]:
            distance = float(distance_case["distance"])
            expected_values = [float(value) for value in distance_case["expected"]]
            assert_vector_close(
                f"{case['name']}/ladder_distance",
                mc.ladder_distance(distance, actual),
                expected_values,
            )
            assert_vector_close(
                f"{case['name']}/ladder_values",
                mc.ladder_values(distance, actual),
                expected_values,
            )
            assert_vector_close(
                f"{case['name']}/Ladder.values",
                object_ladder.values(distance),
                expected_values,
            )
        ladders[case["name"]] = actual
    return ladders


def test_ticks() -> None:
    with (ROOT / "fixtures/golden_ticks.csv").open(newline="") as handle:
        for row in csv.DictReader(handle):
            tier = mc.Tier(
                float(row["epsilon"]),
                float(row["delta"]),
                float(row["p"]),
                float(row["epsilon_ref"]),
            )
            actual = mc.tick_distance(float(row["distance"]), tier)
            assert_close(row["name"], actual, float(row["expected"]))


def test_ladders() -> None:
    with (ROOT / "fixtures/golden_ladders.csv").open(newline="") as handle:
        for row in csv.DictReader(handle):
            tiers = []
            epsilon0 = float(row["epsilon0"])
            delta0 = float(row["delta0"])
            ratio = float(row["ratio"])
            for index in range(int(row["tiers"])):
                scale = ratio**index
                tiers.append(
                    mc.Tier(
                        epsilon0 * scale,
                        delta0 * scale,
                        float(row["p"]),
                        float(row["epsilon_ref"]),
                    )
                )
            actual = mc.ladder_distance(float(row["distance"]), tiers)
            expected = [float(value) for value in row["expected"].split(";")]
            assert_vector_close(row["name"], actual, expected)


def test_binding_conformance_fixture() -> None:
    fixture = load_conformance()
    if fixture["metricchrono_schema"] != "binding_conformance.v1":
        raise AssertionError("unexpected conformance fixture schema")
    tolerance = float(fixture["tolerance"])
    ladders = ladder_map(fixture)

    for case in fixture["tick_distance_cases"]:
        tier = tier_from_doc(case["tier"])
        expected = float(case["expected"])
        assert_close(case["name"], mc.tick_distance(case["distance"], tier), expected, tolerance)
        assert_close(
            f"{case['name']}/try_tick_distance",
            mc.try_tick_distance(case["distance"], tier),
            expected,
            tolerance,
        )

    for case in fixture["smooth_tick_distance_cases"]:
        actual = mc.smooth_tick_distance(
            case["distance"],
            tier_from_doc(case["tier"]),
            case["sharpness"],
        )
        assert_close(case["name"], actual, case["expected"], tolerance)

    for case in fixture["smooth_ladder_distance_cases"]:
        actual = mc.smooth_ladder_distance(
            case["distance"],
            ladders[case["ladder"]],
            case["sharpness"],
        )
        assert_vector_close(case["name"], actual, case["expected"], tolerance)

    for case in fixture["adaptive_ladder_distance_cases"]:
        ticks, decision = mc.adaptive_ladder_distance(case["distance"], ladders[case["ladder"]])
        expected = case["expected"]
        assert_vector_close(case["name"], ticks, expected["ticks"], tolerance)
        assert decision.evaluated_tiers == expected["decision"]["evaluated_tiers"]
        assert decision.first_inactive_tier == expected["decision"]["first_inactive_tier"]
        assert decision.stopped_early == expected["decision"]["stopped_early"]

    for case in fixture["weighted_consensus_cases"]:
        actual = mc.weighted_consensus(case["vectors"], case["weights"])
        assert_vector_close(case["name"], actual, case["expected"], tolerance)

    for case in fixture["event_log_cases"]:
        assert_event_log_case(case, tolerance)

    assert_metric_cases(fixture["metric_cases"], tolerance)
    assert_pair_cases(fixture["pair_cases"], ladders, tolerance)

    for case in fixture["normalize_ticks_cases"]:
        actual = mc.normalize_ticks(case["input"], case["mode"])
        assert_vector_close(case["name"], actual, case["expected"], tolerance)

    for case in fixture["carry_rules_cases"]:
        actual = mc.carry_rules(case["epsilons"])
        if actual != case["expected"]:
            raise AssertionError(f"{case['name']}: expected {case['expected']}, got {actual}")

    for case in fixture["promotion_counter_cases"]:
        with mc.PromotionCounter(case["quotas"]) as counter:
            if counter.quotas != case["quotas"]:
                raise AssertionError(f"{case['name']}: quotas mismatch")
            for step in case["steps"]:
                actual = counter.step(step["event_flags"])
                if actual != step["promoted"]:
                    raise AssertionError(f"{case['name']}: promoted mismatch")
                if counter.counters != step["counters"]:
                    raise AssertionError(f"{case['name']}: counters mismatch")
            counter.reset()
            if counter.counters != case["after_reset_counters"]:
                raise AssertionError(f"{case['name']}: reset counters mismatch")

    assert_rejections(fixture["rejections"], fixture["event_log_cases"][0])


def assert_event_log_case(case: dict[str, Any], tolerance: float) -> None:
    with mc.EventLog(case["tier_count"]) as log:
        assert log.is_empty == case["is_empty_before_append"]
        for record in case["append_records"]:
            actual_index = log.append(record["state_id"], record["ticks"])
            if actual_index != record["expected_index"]:
                raise AssertionError(f"{case['name']}: append index mismatch")

        assert len(log) == case["expected_len"]
        assert log.is_empty == case["is_empty_after_append"]
        assert log.tier_count == case["tier_count"]

        for expected in case["records"]:
            actual = log.record(expected["index"])
            if actual.state_id != expected["state_id"]:
                raise AssertionError(f"{case['name']}: state id mismatch")
            assert_vector_close(
                f"{case['name']}/record[{expected['index']}]",
                actual.ticks,
                expected["ticks"],
                tolerance,
            )
        if len(log.records) != case["expected_len"]:
            raise AssertionError(f"{case['name']}: records property length mismatch")

        for expected in case["first_events"]:
            assert log.first_event(expected["tier"]) == expected["expected"]

        for expected in case["next_events"]:
            assert log.next_event(expected["index"], expected["tier"]) == expected["expected"]

        for summary in case["compact_summaries"]:
            actual = log.compact_summary(summary["tier"])
            expected_items = summary["expected"]
            if len(actual) != len(expected_items):
                raise AssertionError(f"{case['name']}: compact summary length mismatch")
            for left, right in zip(actual, expected_items):
                assert left.index == right["index"]
                assert left.state_id == right["state_id"]
                assert_close("compact summary tick", left.tick, right["tick"], tolerance)


def assert_metric_cases(cases: dict[str, Any], tolerance: float) -> None:
    for case in cases["euclidean_distance"]:
        assert_close(
            case["name"],
            mc.euclidean_distance(case["a"], case["b"]),
            case["expected"],
            tolerance,
        )
    for case in cases["absolute_distance"]:
        assert_close(
            case["name"],
            mc.absolute_distance(case["a"], case["b"]),
            case["expected"],
            tolerance,
        )


def assert_pair_cases(
    cases: dict[str, Any],
    ladders: dict[str, list[mc.Tier]],
    tolerance: float,
) -> None:
    for case in cases["tick_pair"]:
        metric = metric_from_name(case["metric"])
        actual = mc.tick_pair(case["a"], case["b"], metric, tier_from_doc(case["tier"]))
        assert_close(case["name"], actual, case["expected"], tolerance)

    for case in cases["ladder_pair"]:
        metric = metric_from_name(case["metric"])
        actual = mc.ladder_pair(case["a"], case["b"], metric, ladders[case["ladder"]])
        assert_vector_close(case["name"], actual, case["expected"], tolerance)


def metric_from_name(name: str) -> Any:
    if name == "euclidean":
        return mc.Euclidean
    if name == "absolute":
        return mc.Absolute
    if name == "max_abs":
        return mc.MetricFn(max_abs_distance)
    raise AssertionError(f"unknown metric {name}")


def max_abs_distance(a: list[float], b: list[float]) -> float:
    if len(a) != len(b):
        return math.nan
    return max(abs(left - right) for left, right in zip(a, b))


def assert_rejections(rejections: dict[str, Any], event_case: dict[str, Any]) -> None:
    for case in rejections["invalid_tiers"]:
        expect_raises(
            case["name"],
            lambda case=case: mc.Tier.builder()
            .epsilon(case["epsilon"])
            .delta(case["delta"])
            .p(case["p"])
            .epsilon_ref(case["epsilon_ref"])
            .build(),
        )

    for case in rejections["unknown_schema_documents"]:
        if case["kind"] == "tier":
            expect_raises(case["name"], lambda case=case: mc.tier_from_schema(case["document"]))
        else:
            raise AssertionError(f"unknown schema rejection kind {case['kind']}")

    with mc.EventLog(event_case["tier_count"]) as log:
        for record in event_case["append_records"]:
            log.append(record["state_id"], record["ticks"])
        for case in rejections["event_log_out_of_range"]:
            operation = case["operation"]
            if operation == "record":
                expect_raises(case["name"], lambda case=case: log.record(case["index"]))
            elif operation == "next_event":
                expect_raises(
                    case["name"],
                    lambda case=case: log.next_event(case["index"], case["tier"]),
                )
            elif operation == "first_event":
                expect_raises(case["name"], lambda case=case: log.first_event(case["tier"]))
            elif operation == "compact_summary":
                expect_raises(case["name"], lambda case=case: log.compact_summary(case["tier"]))
            else:
                raise AssertionError(f"unknown event log rejection operation {operation}")


def expect_raises(name: str, func: Any) -> None:
    try:
        func()
    except Exception:
        return
    raise AssertionError(f"{name}: expected rejection")


def test_schema_round_trip() -> None:
    with (REPO_ROOT / "tests/golden/ladder.v1.json").open() as handle:
        ladder_doc = json.load(handle)
    ladder = mc.ladder_from_schema(ladder_doc)
    assert mc.ladder_distance(1.0, ladder) == [10.0, 4.0, 2.0]
    assert mc.ladder_to_schema(ladder) == ladder_doc

    with (REPO_ROOT / "tests/golden/tick_vector.v1.json").open() as handle:
        tick_doc = json.load(handle)
    ticks = mc.tick_vector_from_schema(tick_doc)
    assert ticks == [10.0, 4.0, 2.0]
    assert mc.tick_vector_to_schema(ticks) == tick_doc

    with (REPO_ROOT / "tests/golden/consensus_result.v1.json").open() as handle:
        consensus_doc = json.load(handle)
    assert mc.consensus_result_from_schema(consensus_doc) == consensus_doc


if __name__ == "__main__":
    test_ticks()
    test_ladders()
    test_binding_conformance_fixture()
    test_schema_round_trip()
